Implement alert component

rudolfs opened the pull request:
This implements a very simple alert notification. To test it run the app with the proxy not running, it should show the network error message in an alert widget. The styleguide has also been updated with the new element.

xla commented on app/src/molecules/MoleculeAlert.re:
```diff
@@ -0,0 +1,105 @@
+open Atom;
+open Particle;
+
+type severity =
```
Uncertain about this and `t` below as alluded to in another comment down below. My initial reaction was that it should be in a place unrelated to the presentation layer, but then again it's used in here mostly.

rudolfs replied:
I moved the `alert` type to the store, but will leave the `severity` type here, as it's mostly used in this module.

xla replied:
See comment in the store.

xla commented on app/src/molecules/MoleculeAlert.re:
```diff
@@ -0,0 +1,105 @@
+  let make = (~children, ~severity, ~onClick=_ => ()) =>
+    <div className={Styles.alert(severity)}>
+      <div className=Styles.icon>
+        {
```
Not specific to this changeset, but one thing about JSX that I find unergonomic is that it invites to freely interleave markup and normal logic. That's why I usually try to avoid more complex inline code and rather assign the result of such switch statements to variables and use them in the markup.

rudolfs replied:
I was wondering about the way you decide what to leave in JSX and what not, this clears things up. Will adapt my code.

rudolfs commented on app/src/store/StoreAlerts.re:
```diff
@@ -0,0 +1,13 @@
+type action =
+  | Show(Molecule.Alert.t)
```
You're right, this belongs to `StoreAlerts`.

rudolfs commented on app/src/pages/PageProjects.re:
```diff
@@ -76,9 +78,14 @@ let make = () => {
-        <div className="error">
-          {"ERROR: " ++ err##message |> React.string}
-        </div>
+        dispatch(
```
It would just overwrite the message in this case.
I committed some changes that allow showing multiple alerts.

rudolfs commented on app/src/pages/PageProjects.re:
```diff
@@ -76,9 +78,14 @@ let make = () => {
+        dispatch(
+          AlertsAction(
+            StoreAlerts.Show(
+              Alert.{severity: Alert.Error, message: err##message},
```
I agree, we should invest some time to figure out how we guide the user in case there are some problems. We should provide the user with a good explanation of what happened and what to do next to resolve the issue. Additionally we should think about what tone we use in the messaging (formal vs informal/friendly).

But I think this should happen at a later stage when we already have nailed down the user flows then we'll be able to see what the actual error cases are and provide the necessary guidance.

rudolfs commented on app/src/store/StoreAlerts.re:
```diff
@@ -0,0 +1,13 @@
+  | Show(Molecule.Alert.t)
+  | Remove;
+
+type state = {latest: option(Molecule.Alert.t)};
```
Added code for multiple simultaneous alerts.

rudolfs commented:
>Outcome looks awesome.
Yay, thanks for the feedback!

> Currentoy the user has to explicit close them and they don't resolve automatically.
Added support for multiple simultaneous alerts. They're automatically dismissed after 3 seconds now.

> Additionally what happens when multiple errors need to appear and as commented inline can we improve the ergonomics of our errors and how actionable they are.

The designs that @juliendonck provided don't really show how to implement this visually yet, but I added a basic implementation of what multiple simultaneous alerts might look like. I think we can tweak the UX in another PR as this would be a bit out of scope here.

I also added a notification for when the user joins the network.

xla commented on app/src/store/StoreAlerts.re:
```diff
@@ -0,0 +1,30 @@
+open Molecule.Alert;
```
The reason why we have to import this is for severity only, right? For how I understand the dependency graph the store shouldn't know about the high-up layers (anything visual here).

xla commented on app/src/store/StoreAlerts.re:
```diff
@@ -0,0 +1,30 @@
+  | Show(alert)
+  | Remove(alert);
+
+type state = {
```
If we get rid of the id handling here, see comment below, we can have state be as simple as: `type state = array(alert)`.

xla commented on app/src/store/StoreAlerts.re:
```diff
@@ -0,0 +1,30 @@
+};
+
+type action =
+  | Show(alert)
```
Would it be more symmetrically to call it `Add`?

xla commented on app/src/store/ThunkSession.re:
```diff
@@ -38,7 +40,12 @@ let createAccount =
-           SessionAction(Created(account)) |> dispatch |> resolve;
+           SessionAction(Created(account)) |> dispatch;
+           StoreMiddleware.Thunk(
+             ThunkAlerts.showAlert(Alert.Success, "Welcome " ++ keyName),
```
Cute!

xla commented on app/src/store/ThunkAlerts.re:
```diff
@@ -0,0 +1,19 @@
+
+  dispatch(AlertsAction(Show(alert)));
+
+  Js.Promise.make((~resolve, ~reject as _) =>
```
We don't need this to be a promise, as their is not asynchronous interaction after the alert has been set.

rudolfs replied:
True, copy/pasta accident.

xla commented on app/src/store/StoreAlerts.re:
```diff
@@ -0,0 +1,30 @@
+
+type action =
+  | Show(alert)
+  | Remove(alert);
```
This only depends on the `id` not the full alert, would make the surface smaller.

rudolfs replied:
Ha, was thinking of this before, but wanted the alert system to be 1000% robust, you know, in case the random ids clash ; P

Will try it out!

rudolfs replied:
I like that the interface is the same for all the actions. I can change it to reduce the surface area if you insist.

xla replied:
I don't. :> Smaller interfaces just make it easier to reason about and the added benefit is that the caller side needs to hold on to less data, needs less to fullfill the call.

xla commented on app/src/store/ThunkAlerts.re:
```diff
@@ -0,0 +1,19 @@
+open Source;
+
+let showAlert = (severity, message, dispatch, state, _source: source) => {
+  let alert = {severity, message, id: state.alerts.idCounter};
```
Instead of depending on the store keep the id handling, we can use the Random module from Reason and create it locally. This would simplify things.

rudolfs replied:
Ha, was thinking of this before, but wanted the alert system to be 1000% robust, you know, in case the random ids clash ; P

Will try it out!
